### PR TITLE
[node] Adds a 'typesVersions' entry so that Node definitions can start leveraging new compiler features.

### PR DESCRIPTION
**Pull request by rbuckton**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

### CODE REVIEW

**Comment by Flarna** [comment]
Honestly speaking I'm not a fan of duplicating node definitions. Isn't it possible to move the common part (I expect this is >95%) into one file and include it by the 2.1 and 3.1 version?

**Comment by Flarna** [comment]
Honestly speaking I'm not a fan of duplicating node definitions. Isn't it possible to move the common part (I expect this is >95%) into one file and include it by the 2.1 and 3.1 version?

**Comment by rbuckton** [comment]
In the long term, yes. While this is supported by the `"typesVersions"` feature, I discovered an issue with `dtslint` that needs to be addressed first, as it currently disallows using `..` from within the `ts3.1` folder.

**Comment by Flarna** [comment]
Would it work to put the common part in `ts3.1` folder until `dtslint` is fixed?

**Comment by rbuckton** [comment]
Possibly, but if I can I will try to narrow down the dtslint issues in the next few days and possibly revise this PR.

**Comment by rbuckton** [comment]
The bug is actually in types-publisher. I have a PR up to address it: https://github.com/Microsoft/types-publisher/pull/520

**Review by Flarna — CHANGES_REQUESTED** [review]
create a common file to be included in 2.1 and 3.1 variants to avoid duplication.

**Comment by Flarna** [comment]
@rbuckton I requested changes to this PR as discussed above. I have seen that your dtslint fix has been merged therefore I expect this should not land as it is.

**Comment by rbuckton** [comment]
@Flarna: This has been updated, would you mind taking another look?

**Comment by Flarna** [comment]
Sure, will take a look.
But I think someone should tell @typescript-bot that a subfolder starting with `ts` is part of the existing type definition. Currently it thinks `this is a new definition` and as a result no reviewer is invited. I think all Node definition maintainers should get the chance to review.

File: types/node/common.d.ts
Comment (Flarna) [inline types/node/common.d.ts]: I would not write NodeJS 10 here as this will be NodeJs 11 soon and most likely forgotten to be adapted.

File: types/node/index.d.ts
Comment (Flarna) [inline types/node/index.d.ts]: I think it's better to write include or add reference to type....
And it's 2.1 to my knowledge not 2.0.

File: types/node/package.json
Comment (Flarna) [inline types/node/package.json]: should this be ">=3.1"?

File: types/node/ts3.1/index.d.ts
Comment (Flarna) [inline types/node/ts3.1/index.d.ts]: should be include/reference not forward definitons

File: types/node/ts3.1/index.d.ts
Comment (Flarna) [inline types/node/ts3.1/index.d.ts]: This should be remove, the correct info is already at first line

File: types/node/ts3.1/tsconfig.json
Comment (Flarna) [inline types/node/ts3.1/tsconfig.json]: Lib `dom` is included to ensure that node + dom is compatible but do we still need `es6` here?

**Comment by Flarna** [comment]
The tests are still duplicated, lso tslint and quite some part of the typescript config but for these two it's not that critical.
I think the repo Readme should also include a guideline for this approach.
It would be also cool if there is some tool to verify that both definitions "match" to avoid that they get async.

File: types/node/package.json
Comment (rbuckton) [reply types/node/package.json]: No. `>=3.1` expands to `>=3.1.0` and excludes prerelease builds. To include those you have to use a full version with a prerelease modifier of `-0`, which includes _all_ 3.1 builds.

File: types/node/ts3.1/tsconfig.json
Comment (rbuckton) [reply types/node/ts3.1/tsconfig.json]: We need `es5` at a minimum.

**Comment by rbuckton** [comment]
I removed the duplicate `ts3.1/node-tests.ts` and verified that tests are still run for `ts3.1` when only using the relative path reference in `ts3.1/tsconfig.json`.

**Review by Flarna — APPROVED** [review]
Looks fine from my point of view. Would be nice if others working on node typings add also there votes.

**Comment by SimonSchick** [comment]
Can we make this 3.2 so we can use bigint?

**Comment by KonanMentor** [comment]
I have a little concern regarding the file structure: `common.d.ts` and `index.d.ts` approach.

We'll have to modify `common.d.ts` each time another module gets improved typings. These changes
a) hard to review
b) potential place for mistakes during copy-paste

Basically, for changes like #27346 we'll have to do the following:
1. Move a module from `common.d.ts` to `index.d.ts`
2. Put updated module to `ts3.n/index.d.ts`
3. _Somehow_ provision the module for all previous `ts.3(n-1)`, `ts.3(n-2)`, `...` typings (it's gone from `common.d.ts`)

I suggest to split `common.d.ts` into multiple files (one file per module) beforehand. So the file structure will look like
```
├── ...
├── http.d.ts
├── fs.d.ts
├── stream.d.ts
├── index.d.ts
├── ts3.1
│   ├── fs.d.ts
│   ├── index.d.ts
│   └── stream.d.ts
└── ts3.2
    ├── fs.d.ts
    └── index.d.ts
```

Here `fs` and `stream` modules were changed for `ts3.1` and only `fs` module was changed for `ts3.2`.

`ts3.2/index.d.ts` may look like
```
/// <reference path="../ts3.1/stream.d.ts" />
/// <reference path="../http.d.ts" />
/// <reference path="../..." />
...
```

---

In most cases we would like to reuse modules from the previous `ts` typings version. So maybe it even makes sense to create "symlink" for each module of `ts3.n` folder.

It will help to avoid confusion referencing different typescript versions (`ts3.1` and `ts2.1` in the example above).

Also it should be useful when e.g. `ts3.1/http.d.ts` has some useful changes that we expect to have in `ts3.2`. To propagate these changes to `ts3.2` we should not forget to change `ts3.2/index.d.ts` to point to `../ts3.1/http.d.ts` instead of `../http.d.ts`.

So "symlink" version of `ts3.2/index.d.ts` may look like
```
/// <reference path="../ts3.1/stream.d.ts" />
/// <reference path="../ts3.1/http.d.ts" />
/// <reference path="../ts3.1/..." />
...
```

Where `ts3.1/http.d.ts` may contain
```
/// <reference path="../http.d.ts" />
```
simply referencing the same module of the previous `ts` version if no changes are required for this module in this particular `ts` version.

@rbuckton, @Flarna, others, what do you think about it?

**Comment by rbuckton** [comment]
@Flarna, @KonanMentor: Given that this PR has become stale and splitting out node is a fairly large change on its own, I've created https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32567 to address the split on its own (without `"typesVersions"` support).

**Comment by rbuckton** [comment]
Closing in favor of #32691.

**Pull request closed without merging**
